src/jdk.crypto.cryptoki/unix/native/libj2pkcs11/p11_md.c

```diff
@@ -158,48 +154,38 @@ JNIEXPORT jobject JNICALL Java_sun_security_pkcs11_wrapper_PKCS11_connect
     }
 #endif
 
-    if (jGetFunctionList != NULL) {
-        getFunctionListStr = (*env)->GetStringUTFChars(env,
-            jGetFunctionList, 0);
-        if (getFunctionListStr == NULL) {
-            goto cleanup;
-        }
-        C_GetFunctionList = (CK_C_GetFunctionList) dlsym(hModule,
-            getFunctionListStr);
-        if ((systemErrorMessage = dlerror()) != NULL){
-            p11ThrowIOException(env, systemErrorMessage);
-            goto cleanup;
-        }
-        if (C_GetFunctionList == NULL) {
-            TRACE1("Connect: No %s func\n", getFunctionListStr);
-            p11ThrowIOException(env, "ERROR: C_GetFunctionList == NULL");
-            goto cleanup;
-        }
-        TRACE1("Connect: Found %s func\n", getFunctionListStr);
-    } else {
-        // if none specified, then we try 3.0 API first before trying 2.40
+    // if none specified, then we try 3.0 API first before trying 2.40
+    if (jGetFunctionList == NULL) {
         C_GetInterface = (CK_C_GetInterface) dlsym(hModule, "C_GetInterface");
-        if ((C_GetInterface != NULL) && (dlerror() == NULL)) {
+        if (C_GetInterface != NULL) {
             TRACE0("Connect: Found C_GetInterface func\n");
             rv = (C_GetInterface)(NULL, NULL, &interface, 0L);
             if (ckAssertReturnValueOK(env, rv) == CK_ASSERT_OK) {
                 goto setModuleData;
             }
```

Comment:
Do we need an `else goto cleanup` here?

Reply:
Not really, the intention is to continue with the C_GetFunctionList (or the method named by "getFunctionListStr").

Reply:
Good point, I will change it.